### PR TITLE
C++: New Query `cpp/comma-before-misleading-indentation`

### PR DESCRIPTION
**Pull request by d10c**
- C++: Initial commit of `cpp/comma-before-missing-indentation`
- C++: Initial `cpp/comma-before-misleading-indentation`

### CODE REVIEW

**Comment by aibaars** [comment]
@d10c Could you test this query against `git/git` to see if there are any false positives. They use the comma operator quite a bit, but mostly in `for` loop headers.

**Comment by MathiasVP** [comment]
> @d10c Could you test this query against `git/git` to see if there are any false positives. They use the comma operator quite a bit, but mostly in `for` loop headers.

`git/git` is part of the C/C++ DCA suite, so I definitely expect that we'll investigate results on that project before we merge this PR.

File: cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.ql
Comment (MathiasVP) [inline cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.ql:16]: It's indeed a good idea to filter out these results! We actually have a [library predicate](https://github.com/github/codeql/blob/main/cpp/ql/lib/semmle/code/cpp/commons/Exclusions.qll#L126) that handles this for you. I think we should make use of that here.

File: cpp/ql/test/query-tests/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation/test.cpp
Comment (dscho) [inline cpp/ql/test/query-tests/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation/test.cpp:18]: Maybe throw in something like this, for good measure?

```c
for (int i = 0,
     j = 1; i < 10; i += 2,
    j++)
        (void)i;
```

File: cpp/ql/test/query-tests/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation/test.cpp
Comment (geoffw0) [inline cpp/ql/test/query-tests/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation/test.cpp:98]: Great tests / real world testing!

With a query like this I would be happy to simply not report dubious cases such as these three, and in general I wouldn't worry about missed results of any kind too much.

Comment (d10c) [reply cpp/ql/test/query-tests/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation/test.cpp:98]: Yes, but how would you distinguish between this case on line 97 and the one on [line 24](https://github.com/github/codeql/pull/10550/files/0e9b77e7c314509beea26479cb10b47a4044d5e0..cacf78838c61bb855e7bf212a77606552b2fe7a7#diff-5c2f6639161ed558bb0ba560dc0224ec85b97a24b1b17ed1f31af90b9d81c6e4R24-R26)? Without access to source code text, we cannot filter out the "dubious" cases (mixed tabs and spaces). So we have to alert for both cases, or neither.

File: cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.ql
Comment (MathiasVP) [inline cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.ql:14]: Can you add a QLDoc to this predicate?

File: cpp/ql/test/query-tests/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation/test.cpp
Comment (MathiasVP) [inline cpp/ql/test/query-tests/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation/test.cpp:85]: Ouch. I feel bad for whoever ends up reading code like this. This seems like an obvious case of https://wiki.sei.cmu.edu/confluence/display/cplusplus/EXP52-CPP.+Do+not+rely+on+side+effects+in+unevaluated+operands.

File: cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.ql
Comment (MathiasVP) [inline cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.ql:33]: Could this predicate just be defined as `ce.isUnevaluated()`? That certainly includes the two cases you've given here, along with operands of `noexcept`, `alignof` and `typeid` (which are all surrounded by parentheses).

File: cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.ql
Comment (MathiasVP) [reply cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.ql:15]: Ah! The QLDoc makes it a lot more clear what's going on. Thank you!

Could this predicate be written more declaratively as:
```ql
Expr normalizeExpr(Expr e) {
  result =
    min(Expr child, int column |
      child.getParentWithConversions*() = e and column = child.getLocation().getStartColumn()
    |
      child order by column
    )
}
```
?

File: cpp/ql/test/query-tests/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation/test.cpp
Comment (d10c) [reply cpp/ql/test/query-tests/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation/test.cpp:117]: Ouch indeed. There is a certain popular header-only JSON library that does [something like that](https://github.com/syoyo/tinygltf/blob/9bb5806df4055ac973b970ba5b3e27ce27d98148/json.hpp#L4188-L4188). Kept popping up in the MRVA top 1000.

File: cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.ql
Comment (MathiasVP) [reply cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.ql:31]: cc @d10c on ^.

File: cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.ql
Comment (MathiasVP) [inline cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.ql:18]: We should create an internal issue for this. This is most likely a simple oversight.

File: cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.ql
Comment (MathiasVP) [inline cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.ql:16]: I'd write this if this was a member-predicate, but since this is a nonmember-predicate I'd call out the parameter explicitly.
```suggestion
 * Holds if `te` is an implicit `this`.
```

File: cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.ql
Comment (MathiasVP) [inline cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.ql:48]: Should this predicate also include a `SwitchStmt`'s expression?

File: cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.qhelp
Comment (MathiasVP) [inline cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.qhelp:13]: This is very useful information for anyone looking up why they're getting an alert on this new query (and so it should definitely be in here). But I don't think we have any precedent for how to phrase this in a qhelp file, so I hope the Docs team will be able to advise on this.

**Review by MathiasVP — COMMENTED** [review]
This looks super nice! I really appreciate the amount of testing that has gone into this PR. We should be sure to start a DCA run for this before we merge it (and we might as well do that now while we're waiting for a Docs review).

File: cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.ql
Comment (d10c) [reply cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.ql:48]: Mmm, yes, probably.

File: cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.ql
Comment (d10c) [reply cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.ql:31]: Trying it out, looks promising.

**Review by kornbik — COMMENTED** [review]
Finish your review

**Review by MathiasVP — DISMISSED** [review]
The code LGTM! Let's merge this once the Docs review is in 🎉.

File: cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.qhelp
Comment (felicitymay) [inline cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.qhelp:9]: I'm not entirely sure what we're trying to say here. Is this only a problem when we have a comma and bad indentation together? Or, is a comma always a problem, but we're only flagging code where the indentation suggests that it is the comma that's wrong, rather than the indentation?

File: cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.qhelp
Comment (felicitymay) [reply cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.qhelp:13]: A search of the repository revealed that `<warning>` is mostly used in older code, but that we are still using it. Judging by: https://codeql.github.com/codeql-query-help/cpp/cpp-missing-check-scanf/, we don't currently have any styling set up for this warning, so we might want to open an issue to add styling for warnings.

Suggest maybe:

```suggestion
This query has medium precision because CodeQL currently does not distinguish between tabs and spaces in whitespace.
If a file contains mixed tabs and spaces, alerts may highlight code that is correctly indented for one value of tab size but not for other tab sizes.
```

File: cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.qhelp
Comment (felicitymay) [inline cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.qhelp:28]: We usually expect there to be some introduction or explanation of the example. One possibility is to convert the example into three different example files. Alternatively, something like this might work:

```suggestion
<example>
<p>
This example shows three different ways of writing the same code. The first example contains a comma instead of a semicolon which means that the final line is part of the <code>if</code> statement, even though the indentation suggests that it is intended to be separate. The second example looks different put is functionally the same as the first example. It is more likely that the developer intended to write the third example.
</p>
<sample src="CommaBeforeMisleadingIndentation.cpp" />
```

File: cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.qhelp
Comment (felicitymay) [inline cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.qhelp:20]: Would this work?

```suggestion
To ensure that your code is easy to read and review, use standard indentation around the comma operator. Always begin the right-hand-side operand at the same level of
indentation (column number) as the left-hand-side operand. This makes it easier for other developers to see the intended behavior of your code.
```

File: cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.qhelp
Comment (felicitymay) [inline cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.qhelp:23]: My understanding is that most developers prefer to avoid mixed tabs and spaces. Should this be more like:

```suggestion
Use whitespace consistently to communicate your coding intentions. Where possible, avoid mixing tabs and spaces within a file. If you need to mix them, use them consistently.
```

File: cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.qhelp
Comment (felicitymay) [inline cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.qhelp:30]: Are there any articles or guidelines that we could refer to? I had a quick look but couldn't find anything.

File: cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.ql
Comment (felicitymay) [inline cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.ql:3]: We usually aim to structure the description as "syntax X causes problem Y". It's more difficult with non-security queries. Would this work?

```suggestion
 * @description If expressions before and after a comma operator use different indentation, it is easy to misread the purpose of the code.
```

File: cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.ql
Comment (felicitymay) [inline cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.ql:53]: Was there any discussion over putting "for some tab sizes)" in parentheses? I'm not sure that they are needed.

File: cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.cpp
Comment (felicitymay) [inline cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.cpp:2]: Suggest:

```suggestion
 * In this example, the developer intended to use a semicolon but accidentally used a comma:
```

File: cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.cpp
Comment (felicitymay) [inline cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.cpp:13]: Maybe:

```suggestion
 * The use of a comma means that the first example is equivalent to this second example:
```

File: cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.cpp
Comment (felicitymay) [inline cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.cpp:24]:
```suggestion
 * The indentation of the first example suggests that the developer probably intended the following code:
```

File: cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.qhelp
Comment (geoffw0) [reply cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.qhelp:9]: The issue is about bad indentation following a comma, so it requires both of those elements.  Perhaps the confusing part here is actually the use of "left" and "right" to describe relative positions when at least one of these expressions occurs on a different line (with indentation).
```suggestion
If the expression after the comma operator starts at an earlier column than the expression before the comma, then
this suspicious indentation possibly indicates a logic error, caused by a typo that may escape visual inspection.
```

File: cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.qhelp
Comment (geoffw0) [reply cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.qhelp:28]: `put` -> `but` in your suggestion :)

File: cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.ql
Comment (d10c) [reply cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.ql]: Removed the parentheses in a later commit.

**Comment by d10c** [comment]
Thanks for the Doc review @felicitymay! I've made changes based on your suggestions. This PR is now ready for a final review/approval.

File: cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.qhelp
Comment (felicitymay) [reply cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.qhelp]: Thanks Geoffrey 🤦🏻‍♀️

File: cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.qhelp
Comment (d10c) [reply cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.qhelp:33]: Not very targeted, but I added references to Wikipedia.

**Pull request merged**
